### PR TITLE
[FLINK-24227][connectors] Improved AsyncSink base by adding a builder…

### PR DESCRIPTION
**Pull request by CrynetLogistics**
…, adding more commonly used metrics, more java docs around certain implementation tips, and a fatalException callback

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

  - A builder for the AsyncSinkBase
  - Commonly used metrics for the sink
  - Added more javadocs
  - FatalException callback

## Verifying this change

This change added tests and can be verified as follows:

  - Added tests to suite

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (/ no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes /)
  - The serializers: (yes / /)
  - The runtime per-record code paths (performance sensitive): (yes / /)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (/ no /)
  - The S3 file system connector: (/ no /)

## Documentation

  - Does this pull request introduce a new feature? (/ no)
  - If yes, how is the feature documented? (/ JavaDocs /)

### CODE REVIEW

**Comment by CrynetLogistics** [comment]
This PR is a subset of changes for FLINK-17345

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java:58]: Why are these protected rather than private?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java:44]: Why are these protected rather than private?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:143]: I am still not 100% on the concrete part handling the async calls, rather than the base handling threading. Then the concrete implementation just performing simple sync calls. For instance, in KDS sink we are using an Async Kinesis client, which is equivalent to a thread pool here and Sync Kinesis client. Please remind me of the benefit/reason of delegating async to the concrete implementation?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:237]: Is this thread safe when there are multiple in-flight requests? How do we know `ackTime` and `lastSendTimestamp` are the same request?

Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:237]: Is this the same as `lastSendDuration`?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java:58]: I would have thought the concrete implementation of the sink would need to access these values when creating the `SinkWriter`? e.g. here flink-connectors/flink-connector-aws/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSink.java on line 111

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java:44]: Wouldn't the concrete builder need to verify & use these like here `flink-connectors/flink-connector-aws/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkBuilder.java`

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBase.java:58]: Generally speaking I would be more inclined towards private variables with protected getters in this case. It is more maintainable for the future

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java:44]: Same as above

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java]: It's true, race conditions between the in-flight requests occurs. I've modified it such that updates to `ackTime` and `lastSendTimestamp` occur in the same single threaded block now. `lastSendDuration` has been removed since it's not necessary. Added a test to prove too.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:140]: As you mentioned in chat, this design is from flip-171 - idea is to have the concrete sink implementer to pull in an async client. (even though people will have the option to add a thread pool there and have a sync client if they so desire)

**Comment by CrynetLogistics** [comment]
👍

**Comment by CrynetLogistics** [comment]
@fapaul Hi Fabian, would you mind taking a look at this PR which is a subset of the other main PR which is quite big now? 

This would unblock @Yuri Gusev on his task too.

File: flink-connectors/flink-connector-base/src/main/resources/log4j2.properties
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/resources/log4j2.properties:19]: Why did you change this?

Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/resources/log4j2.properties:19]: Fabian had recommended it in #17345. :-)

**Review by dannycranmer — APPROVED** [review]

**Comment by dannycranmer** [comment]
LGTM, I will merge once @fapaul has provided feedback

**Comment by CrynetLogistics** [comment]
Thank you @dannycranmer for your time in providing the review here and in #17345. 👏

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java:908]: We have a `MetricListener` in Flink to ease testing of emitted metrics you take a look at https://github.com/apache/flink/blob/dabc1a894fd5bca1882a071b5d4a015f2d341a32/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java#L153 how to use it.

I think by using it you can get rid of all the metric related mock classes.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java:841]: private?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:301]: Can we make this an instance variable and only instantiate it once? AFAICT it does not hold a reference to a variable of this method?

**Review by fapaul — COMMENTED** [review]
@dannycranmer great idea to split the PRs! It makes it is now much easier to review. I left a couple of minor suggestions. Overall it looks almost ready.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTest.java]: Thanks, it's allowed me to remove 2 classes.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (fapaul) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java]: If it is a field we can probably also expose it to subclasses directly and do not need to change `submitRequestEntries` wdyt?

Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java]: Ah nice, that's even cooler.

Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java]: Done :D

**Comment by CrynetLogistics** [comment]
Please squash the commits when merging.

**Comment by CrynetLogistics** [comment]
@fapaul Let me know if you think the changes are good. Thank you!

**Comment by CrynetLogistics** [comment]
Tests pass 👍

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (fapaul) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:199]: Nit: I'd favor having a protected getter over the direct field access.

**Review by fapaul — APPROVED** [review]
I left one last comment but overall I'll it looks ready to merge 👍

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:199]: I agree with this, since it would be a non-backwards compatible change to reduce scope once merged, let's fix this before merging

Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java]: Thank you! I've also moved this to the top of the class with the other fields.

**Comment by CrynetLogistics** [comment]
Thank you @fapaul for your time reviewing the change.

**Comment by dannycranmer** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
I thought cbe908a has already run? 

> @flinkbot run azure

**Comment by dannycranmer** [comment]
> I thought [cbe908a](https://github.com/apache/flink/commit/cbe908acb91f17e4f485d27ec1a04bcf3eb68c86) has already run?
> 
> > @flinkbot run azure

I see that now, earlier it was showing a failure. Odd

**Pull request merged**
